Add PHP 8.1 prerelease support

rfay opened the pull request:
## The New Solution/Problem/Issue/Bug:

PHP 8.1 prerelease support

## How this PR Solves The Problem:

Uses an external php container with 8.1.0alpha1 in it.

## Manual Testing Instructions:

Try out the instructions in the README. Use a phpinfo.php to see what's happening.

## Related Issue Link(s):

longwave commented:
I tried this out on a Drupal 9 install and while the container did come up, I had to manually install pdo_mysql and then reload php-fpm inside the container:

<?php
docker-php-ext-configure pdo_mysql
docker-php-ext-install pdo_mysql
kill -USR2 1
?>

This could maybe be done with a Dockerfile/build step in the docker-compose config?

I also had to comment out this block in settings.ddev.php as I guess the environment vars aren't quite right:
<?php
if (empty(getenv('DDEV_PHP_VERSION') && getenv('IS_DDEV_PROJECT') == 'true')) {
?>

That was enough to get Drupal 9 to start producing internal errors because of incompatibility issues!

rfay commented:
Outstanding, I'll add that as a Dockerfile.

heddn commented on docker-compose-services/php8_1/README.md:
```diff
@@ -0,0 +1,29 @@
+# PHP8.1 (while in alpha/beta/etc)
+
+DDEV-Local will directly support PHP8 as soon as possible, but not likely until the full release in late 2020. At that time this recipe will be obsolete.
```
```suggestion
DDEV-Local will directly support PHP8 as soon as possible, but not likely until the full release in late 2021. At that time this recipe will be obsolete.
```

heddn commented on docker-compose-services/php8_1/README.md:
```diff
@@ -0,0 +1,29 @@
+
+DDEV-Local will directly support PHP8 as soon as possible, but not likely until the full release in late 2020. At that time this recipe will be obsolete.
+
+In the meantime, you can use the [official docker php image](https://hub.docker.com/_/php) (currently php:8.0.0alpha3-fpm-buster) to serve PHP.  This can be updated as new versions come out.
```
```suggestion
In the meantime, you can use the [official docker php image](https://hub.docker.com/_/php) (currently php:8.1.0alpha1-fpm-buster) to serve PHP.  This can be updated as new versions come out.
```

heddn commented on docker-compose-services/php8_1/docker-compose.php8_1.yaml:
```diff
@@ -0,0 +1,29 @@
+services:
+  php:
+    container_name: ddev-${DDEV_SITENAME}-php
+    image: php:8.1.0alpha1-fpm-buster
```
We might want to wait for alpha2 to be released so opcache is available. See https://bugs.php.net/bug.php?id=81133

rfay replied:
Supposed to be out tomorrow I think. Thanks for all your work on this!

rfay commented:
I added to this significantly to add gd, fileinfo, curl, opcache, pdo_mysql, and updated the README. It also now installs composer to start with. Easy to add other modules in the Dockerfile.

longwave commented on docker-compose-services/php8_1/php-build/usr/local/etc/php/php.ini:
I'm not sure these are needed. This gives errors for me:
```
$ ddev -s php exec php -m
PHP Warning:  PHP Startup: Unable to load dynamic library 'curl' (tried: /usr/local/lib/php/extensions/no-debug-non-zts-20201009/curl (/usr/local/lib/php/extensions/no-debug-non-zts-20201009/curl: cannot open shared object file: No such file or directory), /usr/local/lib/php/extensions/no-debug-non-zts-20201009/curl.so (/usr/local/lib/php/extensions/no-debug-non-zts-20201009/curl.so: cannot open shared object file: No such file or directory)) in Unknown on line 0

Warning: PHP Startup: Unable to load dynamic library 'curl' (tried: /usr/local/lib/php/extensions/no-debug-non-zts-20201009/curl (/usr/local/lib/php/extensions/no-debug-non-zts-20201009/curl: cannot open shared object file: No such file or directory), /usr/local/lib/php/extensions/no-debug-non-zts-20201009/curl.so (/usr/local/lib/php/extensions/no-debug-non-zts-20201009/curl.so: cannot open shared object file: No such file or directory)) in Unknown on line 0
PHP Warning:  Module "fileinfo" is already loaded in Unknown on line 0

Warning: Module "fileinfo" is already loaded in Unknown on line 0
PHP Warning:  Module "gd" is already loaded in Unknown on line 0

Warning: Module "gd" is already loaded in Unknown on line 0
...
```
If I comment these out then the extensions are still loaded:
```
$ ddev -s php ssh
# vi /usr/local/etc/php/php.ini	
# php -m
[PHP Modules]
Core
ctype
curl
date
dom
fileinfo
filter
ftp
gd
...
```

rfay replied:
Removed those enables in the php.ini in https://github.com/drud/ddev-contrib/pull/141/commits/3d6d0e17278d216577706633c855f4670d75e41c - thanks @longwave

longwave commented:
The `php` service in docker-compose also needs
```
links:
  - db
```
This is important if you have multiple ddev projects running at once, otherwise this PHP container (randomly?) talks to the wrong database instance sometimes, which causes all sorts of weird and wonderful error messages.

rfay commented:
Thanks @longwave - added the `links` stanza.

rfay commented:
8.1.0alpha2 is still not in dockerhub, https://hub.docker.com/_/php?tab=tags&page=1&ordering=last_updated&name=8.1.0alpha2

heddn commented:
https://github.com/docker-library/php/issues/1170

rfay commented:
Just tried this with alpha2, and d9 actually ran. I was able to log in even.
